### ISSUE TITLE
TCP Fast Open implementation?

### ISSUE THREAD

**Issue by icylogic** [dotnet/corefx#16375]
#### TCP Fast Open on Windows
Windows has introduced TCP Fast Open in the Anniversary Update serveral months ago. ([Announcing: New Transport Advancements in the Anniversary Update for Windows 10 and Windows Server 2016](https://blogs.technet.microsoft.com/networking/2016/07/18/announcing-new-transport-advancements-in-the-anniversary-update-for-windows-10-and-windows-server-2016/))
> TFO is one full Round Trip Time (RTT) faster than the standard TCP setup which requires a three way-handshake. This leads to latency savings and is very relevant to short web transfers over the Internet where the average latency is on the order of 40 msec.

#### Native API for TCP Fast Open 
This feature could be turned on using native Windows Socket 2 API: [IPPROTO_TCP Socket Options](https://msdn.microsoft.com/en-us/library/windows/desktop/ms738596(v=vs.85).aspx). But it has not been implemented in dotnet ([SocketOptionName.cs](https://github.com/dotnet/corefx/blob/bffef76f6af208e2042a2f27bc081ee908bb390b/src/System.Net.Sockets/src/System/Net/Sockets/SocketOptionName.cs)). 

#### dotnet API relating to TCP Fast Open
According to the Native API Documentation, in addition to passing a new enum value to  the method `setSocketOption` which then calls the native `setsockopt`, we need to set up the buffer including the data to send during the initial connection:
>Note that to make use of fast opens, you should use ConnectEx to make the initial connection, and specify data in that function's lpSendBuffer parameter to be transferred during the handshake process. Some of the data in lpSendBuffer will be transferred under the Fast Open protocol.

The native method `ConnectEx` is called by  [`DoOperationConnect`](https://github.com/dotnet/corefx/blob/master/src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs#L235) which is called by [`socket.ConnectAsync`](https://github.com/dotnet/corefx/blob/master/src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs#L235) on Windows. And the buffer could be set as a property of SocketAsyncEventArgs.

#### dotnet implementation
The most straightforward way to implement this feature seems to be adding a new value to the SocketOptionName enum, which enables this feature only on Windows, but after reading this issue https://github.com/dotnet/corefx/issues/14237, I'm not sure if it's the best way: it seems that members are interested in introducing a new cross-platform API for TCP options.
>It would be nice if the Public API contained all of the possible socket options across Windows, Linux, and Darwin.

PS. TCP Fast Open is natively supported on Linux and OSX. Here's how to use it on Linux: [TCP Fast Open: expediting web services](https://lwn.net/Articles/508865/)

More about TCP Fast Open: [RFC7413](https://tools.ietf.org/html/rfc7413)

**Comment by geoffkizer** [comment]
I agree, we should do this.

**Comment by karelz** [comment]
Next step: We need [formal API proposal](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md)

**Comment by karelz** [comment]
Just edit this one. I often keep the original text in "Original Proposal" section or "Details" as GH doesn't version issue text.
Ping the issue whenever you update top post, summarizing the changes - it is easier to keep track of what happened.

**Comment by icylogic** [comment]
Added a proposal including a short introduction, an example and some problems while leaving the original text unchanged in the section "Original Proposal"

**Comment by karelz** [comment]
Sounds great, thank you!

Re: Compatibility: What happens when you pass the flag on lower Windows version or Linux/OSX version which does not support it? Is it no op? Or failure?

@geoffkizer @davidsh @CIPop @Priya91 any concerns/thumbs up? If I get 2 thumbs up, I can mark it as ready for API review on Tue.

**Comment by geoffkizer** [comment]
I don't think we prevent you from passing arbitrary ints to SetSocketOption.  So something like this should work today:

`            s.SetSocketOption(SocketOptionLevel.Tcp, (SocketOptionName)15, true);`

**Comment by icylogic** [comment]
@geoffkizer Yes, it does work for Windows as I stated in the original proposal. So for Windows after Win10 1607, adding one line to the enum to formalize this API is enough. I'm just not sure what will happen for Unix and Windows before 1607.

**Comment by geoffkizer** [comment]
My point is that we don't need to make any product change for users to do this today.

It might be nice to add the SocketOptionName for this, but that raises the question of whether and how we want to expose OS-specific socket options in SocketOptionName.

**Comment by icylogic** [comment]
OK, if we don't make any change, we can still use this feature on recent Windows. For programmers developing on other platforms which natively support TFO, they may need to wait for this issue https://github.com/dotnet/corefx/issues/14237#issuecomment-266035007 to be solved to use these socket options. It seems that corefx team is interested in introducing a new set of API.
Another issue is about other System.Net API, like WebClient or SslStream, which provide TLS connections and could obviously use this feature. If we don't want make any change to these API using this unexposed feature today, we could wait for TLS 1.3 to come out and make changes together.

**Comment by geoffkizer** [comment]
I think we could certainly consider other APIs, e.g. some sort of way to enable this in a platform-neutral way.

I'm reacting to the "Proposed API" above, which proposes to add a Windows-specific enum value for FastOpen.  My point is that this is a convenience, not a necessity, since you can achieve the desired functionality by just passing (SocketOptionName)15.

**Comment by icylogic** [comment]
That makes sense. It's just the most direct way came to me but actually not a good way.
Maybe we could expose some other APIs that does not take a platform-specific value and translate it in the PAL(Platform Adaptation Layer) like [this](https://github.com/dotnet/corefx/blob/master/src/Native/Unix/System.Native/pal_networking.cpp#L2103-L2113)

**Comment by stephentoub** [comment]
I think we should do two things:
- Add the enum value and appropriately implement it on both Windows and Unix. I'm not sure why the enum value is described as being specific to Windiws... on Unix we'd translate it in the PAL into the right platform value as we do with all other options.
- Add the ability to pass option names in a platform-specific way so that we have a workaround available in the future for other such options. While on Windows it would work to just cast the number to the enum, on Unix we go through a translation layer, so we'd either misinterpret the value if it had an existing match in the enum, or we'd throw for it if the value wasn't known, neither of which is right. We've had a discussion of this in another issue, but I'm on my phone and can't easily find it at the moment. Basically you want the ability to say "pass this exact value to the OS, don't try to interpret/translate it", and leave it up to the app to do the platform detection and determine the right value to use.

**Comment by geoffkizer** [comment]
> Add the enum value and appropriately implement it on both Windows and Unix. I'm not sure why the enum value is described as being specific to Windiws... on Unix we'd translate it in the PAL into the right platform value as we do with all other options.

Is this something we want to continue to do, though?  I realize we did this for compat initially, but it seems like we could get into trouble here.  Would we be better off adding a "bool FastOpen" property on Socket, not unlike NoDelay?

+1 to having a way to do platform-specific options names.

**Comment by stephentoub** [comment]
> but it seems like we could get into trouble here

Can you elaborate?

**Comment by geoffkizer** [comment]
Thinking about it more, I guess there's no actual trouble here; SetSocketOption is already essentially platform-agnostic and translated as appropriate for each platform. 

I guess the interesting question is: Is exposing stuff like this through SetSocketOptions the right API model?  We did this historically because it did in fact map directly to the Windows socket options.  Now that that's not true for Linux/etc, perhaps we should be moving towards a model where these things are exposed directly as properties.  At the very least, it might alleviate potential confusion around exactly what SetSocketOption does.

**Comment by geoffkizer** [comment]
> At the very least, it might alleviate potential confusion around exactly what SetSocketOption does.

Particularly if we're adding something like "SetRawSocketOption" for platform-specific socket option acces.

It would be nice to say:
(a) Platform-agnostic settings are exposed via easy to discover properties; they often map to socket options, but due to differences between platforms, we map these as appropriate
(b) Platform-specific socket options are accessible through a new API(s); use at your own risk
(c) SetSocketOptions is deprecated; use the properties from (a) instead

Thoughts?

**Comment by stephentoub** [comment]
Achieving (c) would mean adding ~30 new members to Socket just for the existing named options, and then additional ones for additional names?

For (a), are there options that are relevant at multiple levels?  If so, those would need methods to take a level in addition to the value to set, along with a corresponding get method?

I actually don't have a strong opinion about it, but the existing mechanism seems sufficient to me for the general case, and maps roughly to the APIs used on both Windows and Unix (get/setsockopt).  The idea of having all of the options available via the Get/SetSocketOption methods and then the most important ones for usability exposed individually as members makes some logical sense to me. My $.02.

Regardless, SetRawSocketOption(int, ...) sounds good to me as the additional platform-specific-numerical-code-you-better-know-what-you're-doing-and-handle-platform-differences-yourself API.

**Comment by geoffkizer** [comment]
Yeah, that does seem like a lot of new members.  

That said, I'm not sure why it's better to expose these via an enum and a generic API.  Is it because the visibility of these things is reduced?  Does enum + generic API imply "here be dragons"?

I realize that the value of doing this is low, given that we already have the enum and generic API.  We should add new stuff like FastOpen using the existing mechanism.  I'm just trying to understand the principles here.

**Comment by icylogic** [comment]
>While on Windows it would work to just cast the number to the enum, on Unix we go through a translation layer, so we'd either misinterpret the value if it had an existing match in the enum, or we'd throw for it if the value wasn't known, neither of which is right.  We've had a discussion of this in another issue, but I'm on my phone and can't easily find it at the moment.

@stephentoub  Are you talking about [this issue](https://github.com/dotnet/corefx/issues/14237#issuecomment-265926068)? This is an example that we can't use some natively-supported socket options on Unix for now. Although we can do this on Windows.

I've mentioned this issue in the original proposal, since I'm not sure whether the implementation of this feature should follow the current workaround. But I think it's the difference of underlying implementation that cause the different behavior of the same API.

If we want to achive two goals on both platforms: 
- Enable developers to use some "unsafe" native API with risks which they know they should take.
- Provide safe-to-use public API which may miss some newest features but are guaranteed to be checked and will work if the check is passed.

For the first one,  SetRawSocketOption() for both Unix and Windows sounds good, and this method would work as what SetSocketOption does for Windows now: pass a raw value to native API.

For the second goal, SetSocketOption on Windows could be modified as: check the option value and generate `System.Net.Sockets.SocketException : Operation not supported` if needed,  just like what TCP_KEEPALIVE would encounter on Unix for now in the above issue. It just don't need the translation. By the way, this could also solve the compability issue of TCP_FASTOPEN on Windows before Win10 1607.
I have to say that the second change would be a breaking change: If one used `SetSocketOption( some-raw-value-for-Windows-only )` for his code ( of course the code maybe only work for Windows), and this raw value is not included in the enum, this would generate an Exception.
If you think the second one is too aggressive, we could focus on SetRawSocketOption for now and I could draft this new proposal.

**Comment by stephentoub** [comment]
I can't tell... are we saying the same thing?  I'm saying:
- socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.FastOpen, true) does what you'd expect: if sets the appropriate option on the underlying socket if supported, or throws an exception if it's not, as is the case with any other option you might try to set that may or may not be supported on a given OS.  How that happens in the code behind this is an implementation detail, and would just follow the established patterns in the codebase for how we do such things today.
- socket.SetRawSocketOption is subject to the interpretation of the current OS, so you better call it with values specific to that OS, e.g. if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) { socket.SetRawSocketOption(15, ...) } else if (RuntimeInformationIsOSPlatform(OSPlatform.Linux) { ... } else { ... } etc.

**Comment by icylogic** [comment]
Sorry, I've checked the source code again and found that although Windows PAL would not check the value before invoking native methods like Unix PAL, it will parse the error code returned from native API and throw exceptions. Then I have no questions about the original SetSocketOption.

I agree that SetRawSocketOption should be used in this way. 

So if we agree on adding SetRawSocketOption, I think the UNIX PAL layer may need to expose a API which does not check the option.
If my understanding is correct, we need either a new method or modify this part:

[SystemNative_SetSockOpt](https://github.com/dotnet/corefx/blob/master/src/Native/Unix/System.Native/pal_networking.cpp#L2211-L2269)
```C#
if (!TryGetPlatformSocketOption(socketOptionLevel, socketOptionName, optLevel, optName))
// TryGetPlatformSocketOption checks options 
// and return false if it could not recognize them.
{
    return PAL_ENOTSUP;
}
```
to something like
```C#
if ((!isRawSocketOption) && (!TryGetPlatformSocketOption(socketOptionLevel, socketOptionName, optLevel, optName)))
// isRawSocketOption is a new parameter for SystemNative_SetSockOpt
{
    return PAL_ENOTSUP;
}
```

And other corresponding changes in `SocketPal.Unix.cs` and `Socket.cs` (adding or overloading).

**Comment by stephentoub** [comment]
> So if we agree on adding SetRawSocketOption

New APIs need to follow the API review process outlined here:
https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md
A new issue should be opened for that.  I realize I may have caused confusion by raising it in this thread, but since it was related, I brought it up.  Now that we're on the same page, it'd be good to have a separate issue for it and this issue can continue focusing on how to support TCP Fast Open in a cross-platform manner.

> I think the UNIX PAL may need to expose an API which does not check the option.

Correct.  But just to be clear, that's all implementation detail.  System.Native.so/dylib is all logically private to .NET Core, so adding new functions there is like adding a private member to a class; implementation detail that doesn't need to be discussed as part of public API design.

**Comment by benaadams** [comment]
Added api request for SetRawSocketOption https://github.com/dotnet/corefx/issues/37122
